server/src/main/java/org/elasticsearch/action/support/TransportAction.java

```diff
@@ -86,18 +86,20 @@ public final Task execute(Request request, TaskListener<Response> listener) {
         execute(task, request, new ActionListener<Response>() {
             @Override
             public void onResponse(Response response) {
-                if (task != null) {
+                try {
                     taskManager.unregister(task);
+                } finally {
+                    listener.onResponse(task, response);
```

Comment:
Should we really resolve the listener successfully if we get an error here?

Reply:
I would say so. Also, previously we would miss to notify the listener, while now we do.

Reply:
I'm not saying we shouldn't enforce calling that listener, but figured we should invoke the `inFailure` if `unregister` (and thus the cancellation callback) throws?

Reply:
I don't have any strong preference either way. Any exception during unregister should be considered as a bug. During testing, it will bubble up to the uncaught exception handler, and tests will fail. In production, we will log a warning. Perhaps, for code simplicity, we can use `onResponse` here.

More importantly, can we make this consistent with the `public final Task execute(Request request, ActionListener<Response> listener)` method in this same class which currently does not invoke the listener in a finally block?

Reply:
done.